### ISSUE TITLE
Tool for CI of Apps of Apps

### ISSUE THREAD

**Issue by timbuchwaldt** [argoproj/argo-cd#9937]
# Summary

Using ArgoCD the way we considers it's intended, by using the app of apps pattern, makes it very hard to test manifests before submitting them to the cluster. You either have to specify special values.yaml for testing or figure out a way how the tree of apps renders. 
I propose implementing a recursive renderer for app of apps structures that allows generating all manifests that result from a given starting app to hand those over to a CI-tool that validates them.

# Motivation

Basically: CI is good, but CI is hard with app of apps.

# Proposal

Optimally a tool that understands what ArgoCD understands: Custom renderers and the relation of app of apps. This could then recursively render all manifests from a given starting app, producing YAML output for furth analysis.

**Comment by crenshaw-dev** [comment]
So something like `argocd app manifests APPNAME --local . --resursive`?

**Comment by timbuchwaldt** [comment]
Wondering: Is this something you would consider implementing or accepting a contribution - or would you consider this out of scope for the ArgoCD project itself?

**Comment by crenshaw-dev** [comment]
@timbuchwaldt I'd absolutely review a contribution! I think I'd only add it myself if I had an Intuit-internal use case (which is becoming increasingly likely as we adopt AppSets more).

**Comment by timbuchwaldt** [comment]
Okay, I'll give it a shot!
I'm still trying to grasp the codebase and some details, but here is what I'm thinking, maybe you can point me in a different direction if mine seems off:

- I'd implement a subcommand like `argocd app local-manifests` that refers to a local repository for starters (I have seen some stuff around using Git credentials, but for a first version I'd probably skip this)
- Internally it calls `repository.GenerateManifests` with the appropriate options from a provided Application CR YAML
- It then parses the generated manifests, extracting all Application CRs, and recursively also resolves those
- The collection of all manifests is returned

**Comment by crenshaw-dev** [comment]
> Internally it calls repository.GenerateManifests with the appropriate options from a provided Application CR YAML

I'd have to look at the code to be certain, but this might call external tools (e.g. Helm) on your local machine. That might be fine, just good to be aware that version differences might be relevant.

Eventually, if it's not part of your initial plan, I'd love to push this work onto the API server. It could pull the top-level manifests from the repo-server, render recursively, and then send the finished manifests to the CLI (or UI, or whatever made the request).

**Comment by JSchlarb** [comment]
Hey @timbuchwaldt, is this a thing I can takeover?
I need a similar thing for local  changes.

Would like to cover also the `app diff`
